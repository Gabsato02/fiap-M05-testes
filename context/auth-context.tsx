"use client"

import { createContext, useContext, useState, useEffect, type ReactNode } from "react"
import { mockUsers } from "@/data/mock-data"
import type { User } from "@/types"

interface AuthContextType {
  user: User | null
  login: (username: string, password: string) => boolean
  register: (username: string, password: string) => boolean
  logout: () => void
}

const AuthContext = createContext<AuthContextType | undefined>(undefined)

export function AuthProvider({ children }: { children: ReactNode }) {
  const [user, setUser] = useState<User | null>(null)

  // Load user from localStorage on initial render
  useEffect(() => {
    const storedUser = localStorage.getItem("user")
    if (storedUser) {
      try {
        setUser(JSON.parse(storedUser))
      } catch (error) {
        console.error("Failed to parse stored user:", error)
        localStorage.removeItem("user")
      }
    }
  }, [])

  const login = (username: string, password: string): boolean => {
    const user = mockUsers.find((u) => u.username === username && u.password === password)

    if (user) {
      const { password, ...userWithoutPassword } = user
      setUser(userWithoutPassword)
      localStorage.setItem("user", JSON.stringify(userWithoutPassword))
      return true
    }

    return false
  }

  const register = (username: string, password: string): boolean => {
    // Check if username already exists
    const userExists = mockUsers.some((u) => u.username === username)
    if (userExists) {
      return false
    }

    // In a real app, we would send this to an API
    // For this mock app, we'll just pretend it worked
    const newUser: User = {
      id: `user-${Date.now()}`,
      username,
      password,
    }

    mockUsers.push(newUser)
    return true
  }

  const logout = () => {
    setUser(null)
    localStorage.removeItem("user")
  }

  return <AuthContext.Provider value={{ user, login, register, logout }}>{children}</AuthContext.Provider>
}

export function useAuth() {
  const context = useContext(AuthContext)
  if (context === undefined) {
    throw new Error("useAuth must be used within an AuthProvider")
  }
  return context
}
