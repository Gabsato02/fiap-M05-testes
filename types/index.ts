// Movie Types
export interface Movie {
  id: string
  title: string
  synopsis: string
  year: number
  genre: string[]
  rating: number
  posterUrl: string
  trailerUrl: string
}

// User Types
export interface User {
  id: string
  username: string
  password?: string
}

// Comment Types
export interface Comment {
  id: string
  movieId: string
  userId: string
  username: string
  text: string
  date: string
}
