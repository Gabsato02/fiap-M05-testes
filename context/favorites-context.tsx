"use client"

import { createContext, useContext, useState, useEffect, type ReactNode } from "react"
import { useAuth } from "./auth-context"

interface FavoritesContextType {
  favorites: string[]
  addFavorite: (movieId: string) => void
  removeFavorite: (movieId: string) => void
}

const FavoritesContext = createContext<FavoritesContextType | undefined>(undefined)

export function FavoritesProvider({ children }: { children: ReactNode }) {
  const [favorites, setFavorites] = useState<string[]>([])
  const { user } = useAuth()

  // Load favorites from localStorage when user changes
  useEffect(() => {
    if (user) {
      const storedFavorites = localStorage.getItem(`favorites-${user.id}`)
      if (storedFavorites) {
        try {
          setFavorites(JSON.parse(storedFavorites))
        } catch (error) {
          console.error("Failed to parse stored favorites:", error)
          localStorage.removeItem(`favorites-${user.id}`)
        }
      }
    } else {
      setFavorites([])
    }
  }, [user])

  const addFavorite = (movieId: string) => {
    if (!user) return

    setFavorites((prev) => {
      const newFavorites = [...prev, movieId]
      localStorage.setItem(`favorites-${user.id}`, JSON.stringify(newFavorites))
      return newFavorites
    })
  }

  const removeFavorite = (movieId: string) => {
    if (!user) return

    setFavorites((prev) => {
      const newFavorites = prev.filter((id) => id !== movieId)
      localStorage.setItem(`favorites-${user.id}`, JSON.stringify(newFavorites))
      return newFavorites
    })
  }

  return (
    <FavoritesContext.Provider value={{ favorites, addFavorite, removeFavorite }}>{children}</FavoritesContext.Provider>
  )
}

export function useFavorites() {
  const context = useContext(FavoritesContext)
  if (context === undefined) {
    throw new Error("useFavorites must be used within a FavoritesProvider")
  }
  return context
}
