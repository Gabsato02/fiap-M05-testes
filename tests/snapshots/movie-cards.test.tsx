import MovieCard from "@/components/movie-card";
import { AuthProvider } from "@/context/auth-context";
import { FavoritesProvider } from "@/context/favorites-context";
import { Movie } from "@/types";
import { render } from "@testing-library/react";
import { describe, expect, it } from "vitest";

const movieMock: Movie = {
  id: "movie-1",
  title: "Inception",
  synopsis: "A thief who steals corporate secrets.",
  year: 2010,
  genre: ["Sci-Fi", "Action"],
  rating: 8.8,
  posterUrl: "/placeholder.svg?height=450&width=300&text=Inception",
  trailerUrl: "https://www.youtube.com/embed/YoHD9XEInc0",
};

describe("Movie Card snapshot", () => {
    it("should match the snapshot", () => {
        const { container } = render(
            <AuthProvider>
                <FavoritesProvider>
                    <MovieCard movie={movieMock} />
                </FavoritesProvider>
            </AuthProvider>
        );

        expect(container).toMatchSnapshot();
    })

    it("should match the snapshot when a movie has multiple genres", () => {
        const movieMockWithMultipleGenres: Movie = {
            ...movieMock,
            genre: ["Sci-Fi", "Action", "Adventure", "Drama"],
        };

        const { container } = render(
            <AuthProvider>
                <FavoritesProvider>
                    <MovieCard movie={movieMockWithMultipleGenres} />
                </FavoritesProvider>
            </AuthProvider>
        );

        // Gera um snapshot do componente renderizado e compara com o snapshot salvo
        expect(container).toMatchSnapshot();
    })
})