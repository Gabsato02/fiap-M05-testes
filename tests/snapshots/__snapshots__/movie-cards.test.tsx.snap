// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Movie Card snapshot > should match the snapshot 1`] = `
<div>
  <a
    href="/movies/movie-1"
  >
    <div
      class="rounded-lg border bg-card text-card-foreground shadow-sm overflow-hidden h-full transition-all duration-200 hover:shadow-lg"
    >
      <div
        class="relative aspect-2/3"
      >
        <img
          alt="Inception"
          class="object-cover"
          data-nimg="fill"
          decoding="async"
          loading="lazy"
          src="/placeholder.svg?height=450&width=300&text=Inception"
          style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
        />
        <button
          aria-label="Add to favorites"
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 hover:text-accent-foreground h-10 w-10 absolute top-2 right-2 bg-black/30 hover:bg-black/50 text-white rounded-full p-1.5"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-heart h-5 w-5"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M19 14c1.49-1.46 3-3.21 3-5.5A5.5 5.5 0 0 0 16.5 3c-1.76 0-3 .5-4.5 2-1.5-1.5-2.74-2-4.5-2A5.5 5.5 0 0 0 2 8.5c0 2.3 1.5 4.05 3 5.5l7 7Z"
            />
          </svg>
        </button>
      </div>
      <div
        class="p-4"
      >
        <h3
          class="font-semibold text-lg line-clamp-1"
        >
          Inception
        </h3>
        <div
          class="flex items-center mt-1 mb-2"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-star h-4 w-4 text-yellow-500 fill-yellow-500"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11.525 2.295a.53.53 0 0 1 .95 0l2.31 4.679a2.123 2.123 0 0 0 1.595 1.16l5.166.756a.53.53 0 0 1 .294.904l-3.736 3.638a2.123 2.123 0 0 0-.611 1.878l.882 5.14a.53.53 0 0 1-.771.56l-4.618-2.428a2.122 2.122 0 0 0-1.973 0L6.396 21.01a.53.53 0 0 1-.77-.56l.881-5.139a2.122 2.122 0 0 0-.611-1.879L2.16 9.795a.53.53 0 0 1 .294-.906l5.165-.755a2.122 2.122 0 0 0 1.597-1.16z"
            />
          </svg>
          <span
            class="ml-1 text-sm"
          >
            8.8
            /10
          </span>
          <span
            class="mx-2 text-gray-400"
          >
            •
          </span>
          <span
            class="text-sm text-gray-600 dark:text-gray-400"
          >
            2010
          </span>
        </div>
        <p
          class="text-gray-600 dark:text-gray-400 text-sm line-clamp-2"
        >
          A thief who steals corporate secrets.
        </p>
      </div>
      <div
        class="items-center p-6 px-4 py-3 border-t flex flex-wrap gap-1"
      >
        <div
          class="inline-flex items-center rounded-full border px-2.5 py-0.5 font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 text-foreground text-xs"
        >
          Sci-Fi
        </div>
        <div
          class="inline-flex items-center rounded-full border px-2.5 py-0.5 font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 text-foreground text-xs"
        >
          Action
        </div>
      </div>
    </div>
  </a>
</div>
`;

exports[`Movie Card snapshot > should match the snapshot when a movie has multiple genres 1`] = `
<div>
  <a
    href="/movies/movie-1"
  >
    <div
      class="rounded-lg border bg-card text-card-foreground shadow-sm overflow-hidden h-full transition-all duration-200 hover:shadow-lg"
    >
      <div
        class="relative aspect-2/3"
      >
        <img
          alt="Inception"
          class="object-cover"
          data-nimg="fill"
          decoding="async"
          loading="lazy"
          src="/placeholder.svg?height=450&width=300&text=Inception"
          style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
        />
        <button
          aria-label="Add to favorites"
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 hover:text-accent-foreground h-10 w-10 absolute top-2 right-2 bg-black/30 hover:bg-black/50 text-white rounded-full p-1.5"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-heart h-5 w-5"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M19 14c1.49-1.46 3-3.21 3-5.5A5.5 5.5 0 0 0 16.5 3c-1.76 0-3 .5-4.5 2-1.5-1.5-2.74-2-4.5-2A5.5 5.5 0 0 0 2 8.5c0 2.3 1.5 4.05 3 5.5l7 7Z"
            />
          </svg>
        </button>
      </div>
      <div
        class="p-4"
      >
        <h3
          class="font-semibold text-lg line-clamp-1"
        >
          Inception
        </h3>
        <div
          class="flex items-center mt-1 mb-2"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-star h-4 w-4 text-yellow-500 fill-yellow-500"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11.525 2.295a.53.53 0 0 1 .95 0l2.31 4.679a2.123 2.123 0 0 0 1.595 1.16l5.166.756a.53.53 0 0 1 .294.904l-3.736 3.638a2.123 2.123 0 0 0-.611 1.878l.882 5.14a.53.53 0 0 1-.771.56l-4.618-2.428a2.122 2.122 0 0 0-1.973 0L6.396 21.01a.53.53 0 0 1-.77-.56l.881-5.139a2.122 2.122 0 0 0-.611-1.879L2.16 9.795a.53.53 0 0 1 .294-.906l5.165-.755a2.122 2.122 0 0 0 1.597-1.16z"
            />
          </svg>
          <span
            class="ml-1 text-sm"
          >
            8.8
            /10
          </span>
          <span
            class="mx-2 text-gray-400"
          >
            •
          </span>
          <span
            class="text-sm text-gray-600 dark:text-gray-400"
          >
            2010
          </span>
        </div>
        <p
          class="text-gray-600 dark:text-gray-400 text-sm line-clamp-2"
        >
          A thief who steals corporate secrets.
        </p>
      </div>
      <div
        class="items-center p-6 px-4 py-3 border-t flex flex-wrap gap-1"
      >
        <div
          class="inline-flex items-center rounded-full border px-2.5 py-0.5 font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 text-foreground text-xs"
        >
          Sci-Fi
        </div>
        <div
          class="inline-flex items-center rounded-full border px-2.5 py-0.5 font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 text-foreground text-xs"
        >
          Action
        </div>
        <div
          class="inline-flex items-center rounded-full border px-2.5 py-0.5 font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 text-foreground text-xs"
        >
          +
          2
        </div>
      </div>
    </div>
  </a>
</div>
`;

exports[`Snapshots do componente movie-card > deve corresponder ao snapshot 1`] = `
<div>
  <a
    href="/movies/movie-1"
  >
    <div
      class="rounded-lg border bg-card text-card-foreground shadow-sm overflow-hidden h-full transition-all duration-200 hover:shadow-lg"
    >
      <div
        class="relative aspect-2/3"
      >
        <img
          alt="Inception"
          class="object-cover"
          data-nimg="fill"
          decoding="async"
          loading="lazy"
          src="/placeholder.svg?height=450&width=300&text=Inception"
          style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
        />
        <button
          aria-label="Add to favorites"
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 hover:text-accent-foreground h-10 w-10 absolute top-2 right-2 bg-black/30 hover:bg-black/50 text-white rounded-full p-1.5"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-heart h-5 w-5"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M19 14c1.49-1.46 3-3.21 3-5.5A5.5 5.5 0 0 0 16.5 3c-1.76 0-3 .5-4.5 2-1.5-1.5-2.74-2-4.5-2A5.5 5.5 0 0 0 2 8.5c0 2.3 1.5 4.05 3 5.5l7 7Z"
            />
          </svg>
        </button>
      </div>
      <div
        class="p-4"
      >
        <h3
          class="font-semibold text-lg line-clamp-1"
        >
          Inception
        </h3>
        <div
          class="flex items-center mt-1 mb-2"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-star h-4 w-4 text-yellow-500 fill-yellow-500"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11.525 2.295a.53.53 0 0 1 .95 0l2.31 4.679a2.123 2.123 0 0 0 1.595 1.16l5.166.756a.53.53 0 0 1 .294.904l-3.736 3.638a2.123 2.123 0 0 0-.611 1.878l.882 5.14a.53.53 0 0 1-.771.56l-4.618-2.428a2.122 2.122 0 0 0-1.973 0L6.396 21.01a.53.53 0 0 1-.77-.56l.881-5.139a2.122 2.122 0 0 0-.611-1.879L2.16 9.795a.53.53 0 0 1 .294-.906l5.165-.755a2.122 2.122 0 0 0 1.597-1.16z"
            />
          </svg>
          <span
            class="ml-1 text-sm"
          >
            8.8
            /10
          </span>
          <span
            class="mx-2 text-gray-400"
          >
            •
          </span>
          <span
            class="text-sm text-gray-600 dark:text-gray-400"
          >
            2010
          </span>
        </div>
        <p
          class="text-gray-600 dark:text-gray-400 text-sm line-clamp-2"
        >
          A thief who steals corporate secrets.
        </p>
      </div>
      <div
        class="items-center p-6 px-4 py-3 border-t flex flex-wrap gap-1"
      >
        <div
          class="inline-flex items-center rounded-full border px-2.5 py-0.5 font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 text-foreground text-xs"
        >
          Sci-Fi
        </div>
        <div
          class="inline-flex items-center rounded-full border px-2.5 py-0.5 font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 text-foreground text-xs"
        >
          Action
        </div>
      </div>
    </div>
  </a>
</div>
`;
